appscale/tools/agents/azure_agent.py

```diff
@@ -399,19 +405,34 @@ def create_virtual_machine(self, credentials, network_client, network_id,
       uri='https://{0}.blob.core.windows.net/vhds/{1}.vhd'.
         format(storage_account, vm_network_name))
 
-    image_hd = VirtualHardDisk(uri=parameters[self.PARAM_IMAGE_ID])
     os_type = OperatingSystemTypes.linux
-    os_disk = OSDisk(os_type=os_type, caching=CachingTypes.read_write,
-                     create_option=DiskCreateOptionTypes.from_image,
-                     name=vm_network_name, vhd=virtual_hd, image=image_hd)
-
+    azure_image_id = parameters[self.PARAM_IMAGE_ID]
+
+    # Publisher images are formatted Publisher:Offer:Sku:Tag
+    if re.search(".*:.*:.*:.*", azure_image_id):
+      AppScaleLogger.log("Using publisher image {}".format(azure_image_id))
+      image_ref_params = azure_image_id.split(":")
+      image_ref = ImageReference(publisher=image_ref_params[0],
+                                 offer=image_ref_params[1],
+                                 sku=image_ref_params[2],
+                                 version=image_ref_params[3])
+      os_disk = OSDisk(os_type=os_type, caching=CachingTypes.read_write,
+                       create_option=DiskCreateOptionTypes.from_image,
+                       name=vm_network_name, vhd=virtual_hd)
+      storage_profile = StorageProfile(image_reference=image_ref,
+                                       os_disk=os_disk)
+    else:
+      image_hd = VirtualHardDisk(uri=parameters[self.PARAM_IMAGE_ID])
+      os_disk = OSDisk(os_type=os_type, caching=CachingTypes.read_write,
+                       create_option=DiskCreateOptionTypes.from_image,
+                       name=vm_network_name, vhd=virtual_hd, image=image_hd)
+      storage_profile = StorageProfile(os_disk=os_disk)
```

Comment:
Could you pull out the os_disk and storage_profile from the if condition so that there is less duplication like this: